### PR TITLE
[service] Add a CompilationSession abstract base class.

### PR DESCRIPTION
**Pull request by ChrisCummins**
The CompilationSession class encapsulates an incremental compilation
session.

This is the first in a series of patches that separates the RPC server
from the compilation session.

Issue #254.

### CODE REVIEW

**Comment by ChrisCummins** [comment]
Marking this ready for review but no rush to merge it while we're still discussing the design in #254.

File: compiler_gym/service/compilation_session.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    action_spaces: List[ActionSpace] = []
+    """A list of action spaces describing the capabilities of the compiler."""
+
+    observation_spaces: List[ObservationSpace] = []
+    """A list of feature vectors that this compiler provides."""
+
+    def __init__(
+        self, working_dir: Path, action_space: ActionSpace, benchmark: Benchmark
+    ):
```
Comment (hughleat) [inline compiler_gym/service/compilation_session.py:49]: Just curious: why doesn't this return a `status` like the c++ code?

**Review by hughleat — APPROVED** [review]
If the protobuf stuff is okay to keep then this LGTM :-)

File: compiler_gym/service/compilation_session.py (hunk 1 above)
Comment (ChrisCummins) [reply compiler_gym/service/compilation_session.py:49]: We [don't use exceptions in C++](https://google.github.io/styleguide/cppguide.html#Exceptions) (FWIW I don't have particularly strong opinions on the matter), the `CompilationSession` constructor in C++ is a safe default, and any of the three methods that can fail (init, compute feature, run action) return `Status`.

For Python, exceptions are fine and all of the runtime logic that interacts with `CompilationSession` will catch exceptions and set the relevant RPC error code

**Pull request merged**